### ISSUE TITLE
How important is Axis and QLLM support

### ISSUE THREAD

**Issue by wejoncy** [mobiusml/hqq#9]
Hi HQQ team, Happy new year.
I's a great work and thanks for  the insightful idea of which is really faster during quantization. I installed hqq and ran examples in a minutes.

When I dive into your code and I found there is a parement named "axis" to indicate which direction the compression will go on. I got a question on my mind, would axis=0 be better then axis=1? I just checked that it seems GPTQ and AWQ both used axis=1 by default. However it's 0 in HQQ.

**I tried to set it to 0 and do the quantization, it looks good as well. So, Do you have any comments on the quantization axis?**

Besides, I integrate HQQ into my general quantization [QLLM](https://github.com/wejoncy/QLLM), so it works for all models in huggingface for now. The same time, it could be easily support mix-bits quantization which is naturelly supported by qllm.

**Comment by mobicham** [comment]
Hi @wejoncy ! Thanks for your message and happy new year!

Regarding the axis question, I guess it doesn't matter or maybe for some models some axis would work better, difficult to say. We did all our benchmarks and tests with axis=0 so it would be better to just keep it that way. Since the quantization is very fast, you can even run it with both and choose the one that minimizes the error between the reconstructed and original weights!

That's great! For now, we do it manually for each architecture, but one can just loop over the layers and if it's nn.Linear -> quantize. The reason why we also do it manually is because for some models like Mixtral, giving different quantization settings to different layers works much better, see this thread: https://github.com/mobiusml/hqq/issues/2

**Comment by rainyBJ** [comment]
Hi, @wejoncy, maybe you can reference this paper: https://arxiv.org/pdf/2309.15531.pdf. Quantization along axis=0, may be helpful for limiting the outlier in fewer groups as figure 1 illustrated. In fact, the activation outlier is quite common in different LLM models. 
<img width="650" alt="image" src="https://github.com/mobiusml/hqq/assets/49946568/78dbc03e-d6c3-4192-8351-79db2812383f">
And I did some experiments using HQQ on PPL for OPT 125m->13B using LLM eval harness based on my own implementation for HQQ, the results shows that quantization along axis=0 is constantly better than axis=1. The results(w4 g128 asym):
| OPT | 125M | 1.3B | 2.7B | 6.7B | 13B |
|-------|-------|-------|-------|-------|-------|
|  axis=1 |  39.19 |  18.16 |  15.75 |  13.34 |  11.91 |
|  axis=0 |  36.07 |  17.04 |  14.65 |  12.46 |  11.63 |

**Comment by mobicham** [comment]
Thanks @rainyBJ for the reference! It will be interesting to do the comparison by keeping the quantized weights/scale/zero in float32 and remove the clipping. I noticed some instabilities with fp16 which could contribute to this. 
To stay agnostic to the axis parameter, you can run the quantize step with both and pick the one that minimizes the error for each layer separately in: https://github.com/mobiusml/hqq/blob/master/hqq/core/quantize.py#L294

```Python
@torch.inference_mode()
def quantize(self, W, weight_quant_params, scale_quant_params, zero_quant_params):
	quant_scale = scale_quant_params is not None
	quant_zero  = zero_quant_params  is not None

	self.in_features, self.out_features = W.t().shape

	if('axis' in weight_quant_params): 
		del weight_quant_params['axis']
	
	#Quantize - axis agnostic
	W_q_0 , meta_0 = Quantizer.quantize(W, **weight_quant_params, axis=0) 
	err_0          = torch.abs(W - Quantizer.dequantize(W_q_0, meta_0)).mean()

	W_q_1 , meta_1 = Quantizer.quantize(W, **weight_quant_params, axis=1) 
	err_1          = torch.abs(W - Quantizer.dequantize(W_q_1, meta_1)).mean()

	if(err_0<err_1):
		W_q, meta = W_q_0, meta_0
		del W_q_1 , meta_1
	else:
		W_q, meta = W_q_1, meta_1
		del W_q_0 , meta_0


	meta.update({'quant_scale':quant_scale, 'quant_zero':quant_zero})
	if(meta['quant_scale']):
		meta['scale_q'] , meta['meta_scale'] = Quantizer.quantize(meta['scale'], **scale_quant_params); del meta['scale']
	if(meta['quant_zero']):
		meta['zero_q'], meta['meta_zero']    = Quantizer.quantize(meta['zero'],  **zero_quant_params);  del meta['zero']

	self.W_q   = W_q
	self.meta  = meta 
	self.cuda()
	self.ready = True
```

**Comment by wejoncy** [comment]
Really Thanks @mobicham  and @rainyBJ for your reply and the detail explanation.
It's move obvious for small model to have better PPL with axis=0 from your experiment results. I guess we can also get better results for the other algorithms.

@mobicham  Yeah, It's really amazing for the quantization speed of HQQ, it wouldn't take too long to pick the best axis for different layers. Do you think if we can have an algorithm to select the best paraments, say bits, group-size, axis, and we can have the Pareto optimality between size and PPL.  
May the easiest way be to sort all layers by loss and re-quantize those candidate layers.

**Comment by mobicham** [comment]
Yeah , not all layers are equally sensitive to quantization. For example, for the Mixtral model, the experts can be quantized to 2-bit and the attention layers to 4-bit and there's only an additional 0.20GB compared to using 2-bit for all the layers but the performance is significantly better. So it would make sense to give different parameters to different layers. 

The easiest way is to use `quant_config` and set it for each layer type like in here : https://huggingface.co/mobiuslabsgmbh/Mixtral-8x7B-v0.1-hf-attn-4bit-moe-2bit-HQQ

**Comment by wejoncy** [comment]
Got it. Thank you.

I am thinking if there is a way to do it without human involved.
